### ISSUE TITLE
Lot's of questions (future, memory mgmt, metaprogramming, etc.)

### ISSUE THREAD

**Issue by dumblob** [Swallow-lang/swallow#11]
Please refer to https://github.com/aardappel/lobster/issues/136 , read the thread to understand why I'm asking these questions and please answer them here :wink:.

Additionally I'd like to ask:

a) Why don't you want to join Lobster or Cython or even RPython devs instead of starting from scratch?
b) How do you want to support parallel computation on multiple CPU/GPU/FPGA cores?
   This is by far the most important point utterly needed as explained in https://github.com/vlang/v/discussions/11608 because it affects nearly all design decisions made when designing a language.
   This ranges from "each statement & expression is run in parallel" ([Kind](https://github.com/uwu-tech/Kind ), [ParaSail](https://www.embedded.com/design/programming-languages-and-tools/4428704/2/Alternatives-to-C-C--for-system-programming-in-a-distributed-multicore-world ), [BMDFM](http://www.bmdfm.com/pdf/Dataflow_Multicore_Manycore.pdf ), [DawnCC](https://github.com/gleisonsdm/DawnCC-Compiler ) - actually not a new language but a smart compiler supporting even non-uniform platforms like big.LITTLE, ... ) through channel/pipe-based parallelism ([Dao](https://github.com/daokoder/dao ), [Go](https://golang.org/ ), [V](https://vlang.io/ ), [POSIX sh](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sh.html ), [Til](https://github.com/til-lang/til ), ...) to non-builtin support (i.e. pthreads, actors, [structured concurrency or nurseries](https://en.wikipedia.org/wiki/Structured_concurrency ), etc.).

**Comment by SaptakBhoumik** [comment]
@dumblob Thank you for your questions and here is your answer

> a) Why don't you want to join Lobster or Cython or even RPython devs instead of starting from scratch?

Well, I would like to add some unique features which I cant if I join such projects. So I am creating it from scratch

> b) How do you want to support parallel computation on multiple CPU/GPU/FPGA cores?

Yes it is planned

**Comment by SaptakBhoumik** [comment]
For memory management, my plan is to use auto-free where the compiler will insert free calls wherever necessary

**Comment by SaptakBhoumik** [comment]
Here is a benchmark between swallow and python
![image](https://user-images.githubusercontent.com/74509281/135803290-0c6c5a28-76e8-4098-aa3b-8bed50db8ac7.png)
Here I calculated the Fibonacci series recursively from 0 to 40 and python(left side) took 32secs whereas swallow(right side) took only 1sec thus making it 32 times faster. Also, you can see that its syntax is very similar to that of python. For comparison here is a benchmark with c
![image](https://user-images.githubusercontent.com/74509281/135803371-df72b0cd-130f-488f-8fd8-2e17cee034ce.png)
And as you can see that swallow is as fast as c(both took 1sec)

**Comment by SaptakBhoumik** [comment]
In the future, it aims to be a replacement for python. Since it gets transcompiled to c it can be used for systems programming as well.

**Comment by SaptakBhoumik** [comment]
It will not have macros to keep the language simple

**Comment by dumblob** [comment]
Thanks for quick reactions!

Performance is nice, but there are a dozen other languages with high-level feel offering C speeds. Let's start with contributing Swallow solutions to https://github.com/kostya/benchmarks/ and let everyone see and judge by themself.

### So back to the questions:

1. Where can we find the plan for the foreseeable future of Swallow?
2. Any real world use cases?
3. Any production systems/apps/... using it?
4. Any commercial subjects investing into Swallow ("backing it up")?
5. Could you add Swallow to https://github.com/kostya/benchmarks ?
6. How about FFI? My question is two fold. I'm interested in both static (compile-time) FFI for speed (and safety as a second and less important reason; imagine generating bindings from C header files) and runtime FFI (for convenience; imagine [PythonCall.jl](https://cjdoris.github.io/PythonCall.jl/stable/pycall/ ) - not to be confused wyth PyCall.jl).
7. Does it compile to JS or at least WASM to support web ecosystem deployment?

> > a) Why don't you want to join Lobster or Cython or even RPython devs instead of starting from scratch?
> 
> Well, I would like to add some unique features which I cant if I join such projects. So I am creating it from scratch

Which unique features do you have in mind?

> > b) How do you want to support parallel computation on multiple CPU/GPU/FPGA cores?
> 
> Yes it is planned

How will it look like? What are the concrete goals?

### And some more additional questions:

c) Will you support sub-second compilation of mid-size projects (like V: https://fast.vlang.io/ )?
d) Will the resulting binaries/apps support iOS & Android deployment?
e) Will you have a full ecosystem with seamless development experience (linter, autoformatter, package manager, REPL or console IDE, Ekam-like [auto-recompiler](https://github.com/vlang/v/pull/9577 ), etc.)?
f) Will it be "safe **by default** at all cost"? (except for traditional safety as memory management, null dereferencing, etc. it'll also [not allow the user to use arbitrary real number literal](https://github.com/vlang/v/discussions/11855 ) if it's not exactly representable by the underlying language - e.g. by a binary float of given size, etc.)
g) Will you **guarantee** all `null`/`none`/`nil` values are checked for safety in **compile time**? - see e.g. how it could be easily implemented with sumtypes in https://github.com/vlang/rfcs/pull/7#issuecomment-921706984
h) Will you offer a no-hassle way to **guarantee** "alternative computational branches" - especially branches dealing with error handling - are **always handled**? (any variation on *exceptions* is the worst possible solution - I'd recommend [sumtypes with syntactic sugar for *result* and *optional* cases](https://github.com/vlang/rfcs/pull/7 ))
i) If not (hygienic) macros, then how will you solve exceptional cases where you need to do unsafe operations or where you'd need to scratch the language semantics and do something quite differently (e.g. allocate the given piece of memory on stack, or disable bounds checking, or you name it...)?
j) Will it support "live/reactive" data flow (e.g. [ECS](https://github.com/Pauan/rust-signals/discussions/31 ), pub-sub/signal-slot, etc.)? (necessary for GUIs, any streaming processing, linearly scalable multicore utilization, etc.)
k) Will you resist the artificial urge and **not** implement "async" support and instead offer it as standard library feature to make it truly fully explicit (pretty much like [Julia does](https://docs.julialang.org/en/v1/manual/asynchronous-programming/ ))?

**Comment by SaptakBhoumik** [comment]
@dumblob I would love to add swallow to  https://github.com/kostya/benchmarks/ but this language is still not mature enough to solve all of these problems

> Where can we find the plan for the foreseeable future of Swallow?

Tbh  this language is still very young so I can't predict its future

> Any real world use cases?

It can be used for science, making low-level software, ai etc

> Any production systems/apps/... using it?

As I told you this language is very young for such sort of task

> Any commercial subjects investing into Swallow ("backing it up")? 

No

> How about FFI? My question is two fold. I'm interested in both static (compile-time) FFI for speed (and safety as a second and less important reason; imagine generating bindings from C header files) and runtime FFI (for convenience; imagine PythonCall.jl - not to be confused wyth PyCall.jl).

Well you can call functions from C using Ccode. Here is an example
```
himport "headerfile.h"
cimport "cfile.h"
flag compiler_flag
linux_flag os_specific_compiler_flag
def wrapper():
    Ccode c_function() Ccode
```
You can also use inline assembly like this
```
def main():
    int arg1 = 45
    int arg2 = 50
    int add = 0
    print("It should add 45 and 50 using asm and print it\n")
    asm("addl %%ebx, %%eax;" : "=a" (add) : "a" (arg1) , "b" (arg2))
    printf("%lld", add)
```
> Does it compile to JS or at least WASM to support web ecosystem deployment?

WASM is planned

> Which unique features do you have in mind?

Pointers, constants,auto-free, builtin support for threading and concurrence

> How will it look like? What are the concrete goals?

Some builtin library, decorators etc and wrapper for cuda

>Will you support sub-second compilation of mid-size projects (like V: https://fast.vlang.io/ )

I cant guarantee that  but the resulting binary will be very very fast

>Will the resulting binaries/apps support iOS & Android deployment?

Not sure

>Will you have a full ecosystem with seamless development experience (linter, autoformatter, package manager, REPL or console IDE, Ekam-like auto-recompiler, etc.)

Hopefully 

> Will it be "safe by default at all cost"? (except for traditional safety as memory management, null dereferencing, etc. it'll also not allow the user to use arbitrary real number literal if it's not exactly representable by the underlying language - e.g. by a binary float of given size, etc.)

Well we can only guarantee memory safety rest everything is in the hands of the programmer 

>Will you guarantee all null/none/nil values are checked for safety in compile time? - see e.g. how it could be easily implemented with sumtypes in vlang/rfcs#7 (comment)

No, we can only guarantee memory safety rest everything is in the hands of the programmer 

>h) Will you offer a no-hassle way to guarantee "alternative computational branches" - especially branches dealing with error handling - are always handled? (any variation on exceptions is the worst possible solution - I'd recommend sumtypes with syntactic sugar for result and optional cases)

Yes most probably

>If not (hygienic) macros, then how will you solve exceptional cases where you need to do unsafe operations or where you'd need to scratch the language semantics and do something quite differently (e.g. allocate the given piece of memory on stack, or disable bounds checking, or you name it...)?

Well for performing unsafe operations you don't have to use unsafe block

>Will it support "live/reactive" data flow (e.g. ECS, pub-sub/signal-slot, etc.)? (necessary for GUIs, any streaming processing, linearly scalable multicore utilization, etc.)

Yes hopefully

> Will you resist the artificial urge and not implement "async" support and instead offer it as standard library feature to make it truly fully explicit (pretty much like Julia does)?

It will have builtin support  for "async"

**Comment by dumblob** [comment]
Ok, it seems Swallow is in its true infancy. Let's get back to these questions and reopen this issue a year later :wink:. Good luck so far!
